# This patch adds cacheid to gpg-agent. This is needed for the Mac OS X keychain.

--- a/agent/agent.h
+++ b/agent/agent.h
@@ -276,7 +276,7 @@
                   struct pin_entry_info_s *pininfo);
 int agent_get_passphrase (ctrl_t ctrl, char **retpass,
                           const char *desc, const char *prompt,
-                          const char *errtext, int with_qualitybar);
+                          const char *errtext, int with_qualitybar, const char *cacheid);
 int agent_get_confirmation (ctrl_t ctrl, const char *desc, const char *ok,
 			    const char *notokay, int with_cancel);
 int agent_show_message (ctrl_t ctrl, const char *desc, const char *ok_btn);
--- a/agent/call-pinentry.c
+++ b/agent/call-pinentry.c
@@ -853,7 +853,7 @@ agent_askpin (ctrl_t ctrl,
 int 
 agent_get_passphrase (ctrl_t ctrl,
                       char **retpass, const char *desc, const char *prompt,
-                      const char *errtext, int with_qualitybar)
+                      const char *errtext, int with_qualitybar, const char *cacheid)
 {
 
   int rc;
@@ -869,6 +869,15 @@
   if (rc)
     return rc;
 
+	
+  if (cacheid) {
+    snprintf (line, DIM(line)-1, "OPTION cache-id=%s", cacheid);
+    rc = assuan_transact (entry_ctx, line, NULL, NULL, NULL, NULL, NULL, NULL);
+    if (rc)
+      return unlock_pinentry (rc);
+  }
+	
+	
   if (!prompt)
     prompt = desc && strstr (desc, "PIN")? "PIN": _("Passphrase");
 
--- a/agent/command.c
+++ b/agent/command.c
@@ -1270,7 +1270,7 @@
     next_try:
       rc = agent_get_passphrase (ctrl, &response, desc, prompt,
                                  repeat_errtext? repeat_errtext:errtext,
-                                 opt_qualbar);
+                                 opt_qualbar, cacheid);
       xfree (repeat_errtext);
       repeat_errtext = NULL;
       if (!rc)
@@ -1287,7 +1287,7 @@
               char *response2;
 
               rc = agent_get_passphrase (ctrl, &response2, desc2, prompt,
-                                         errtext, 0);
+                                         errtext, 0, cacheid);
               if (rc)
                 break;
               if (strcmp (response2, response))
